app/src/androidTest/java/org/mozilla/fenix/helpers/TestHelper.kt

```diff
@@ -117,4 +121,28 @@ object TestHelper {
             }
         }
     }
+
+    fun setNetworkConnection(enabled: Boolean) {
+        val mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
+
+        when (enabled) {
+            true -> {
+                mDevice.executeShellCommand("svc data enable")
+                mDevice.executeShellCommand("svc wifi enable")
+
+                // Stopping any next steps until the network is fully connected
+                // Make sure to unregister the Idling resource after using this method
+                IdlingRegistry.getInstance().register(NetworkConnectedIdlingResource())
```

Comment:
Why did you decide to use `IdlingResource` as a way to wait for the data to be disabled/enabled? Is it to avoid blocking the test thread and have a mechanism to time out the test if the action never happens?

With the current implementation, I'm concerned that:
- it's very easy to forget to deregister the `IdlingResource` and this state could propagate to other tests
- keeping the `IdlingResource` alive alive longer than it takes to enable/disable the network might have unexpected side effects on the tests (it's _probably_ okay but it seems hard to think through the cases)
- (see my other comment) the `IdlingResource`s are not actually being unregistered

A few alternative implementations I can think of that might help:
- if the IdlingResource deregistered itself once it was idle (though that makes it less re-usable; this might be my least favorite of these implementations)
- implementing the network calls a `Rule` such that the `Rule` registers and deregisters the IdlingResources and give you access to the `setNetworkConnection` APIs
- removing `IdlingResource` and implementing the wait + time out ourselves inside the method to change the network.

The last one seems the simplest to me without unexpected side effects but perhaps it's more error prone? In which case 2 is better but maybe it has unexpected consequences to keep the IdlingResource around? I'm not sure. It seems like a clever and interesting use case for `IdlingResource` but I want to make sure I actually understand why we're using it and the implications as I haven't used it much. :)

Reply:
Yes, we're using Idling Resources in many of our tests to wait just until something is ready, not more or less. If we'd set a strict timeout it could result in flakiness if it takes longer or we'd have tests wait an unnecessary amount of time. I'll take your second suggestion and move both the register & unregister to the test class.